### ISSUE TITLE
Key listenter setting and issues in SPA

### ISSUE THREAD

**Issue by Mister-Hope** [algolia/docsearch#1363]
Hi, this is Mr.Hope from VuePress Team. 

Docsearch is not providing a method to destory itself or any way to customize (or remove) the search key listener in docs. So I wonder if there is a way, otherwise this will trigger error in SPA application.

Visit https://mrhope.site/ and scroll down then click the first arrticle. When you click `Ctrl + K`, you will find 2 docsearch popup is displayed.

In VuePress, we are using SPA have "layout" comopnent and different page may choose different layout. So when you switch to the first article, the whole page is rerendered because they have different layout, so docsearch is reinitialize. But the previous key listener seems still exist, so keyboard actions seems not correct.

**Comment by Mister-Hope** [comment]
Hi, any update with this one? Can we add a detroy method to clear docsearch event listener and destory the dom?

**Comment by Mister-Hope** [comment]
Hello?

**Comment by shortcuts** [comment]
(Hey there, sorry for the delay, added context here https://github.com/algolia/docsearch/issues/1734)

**Comment by Mister-Hope** [comment]
Sure, I will try to build one, I jsut want to make sure if this sounds ok to you before starting contribution.

**Comment by Mister-Hope** [comment]
Hi, any update about this one? I am still meeting users reporting this frequently.

**Comment by bitspittle** [comment]
Hey, we're new users to Algolia with a SPA framework of our own. Today was the first day we set up the search bar and we ran into this in testing.

The original comment already discusses their repro steps, so you can probably ignore these, but I'll add my own:

**Repro steps:**

- Visit https://preview.kobweb.varabyte.com/docs/
- Press Ctrl K / Cmd K to bring up the search box
- Type "Routing" and press ENTER
- Site navigates, dialog closes ✅
- Press Ctrl K / Cmd K again
- Type "IDE" and press ENTER
- Site navigates, dialog stays open ❌

**Interesting alternate steps that work (using the mouse):**

- Visit https://preview.kobweb.varabyte.com/docs/
- Press Ctrl K / Cmd K to bring up the search box
- Type "Routing" and press ENTER
- Site navigates, dialog closes ✅
- Use the mouse to click on the search button
- Type "IDE" and press ENTER
- Site navigates, dialog closes ✅

**Comment by bitspittle** [comment]
FYI we've fixed this locally using the following commit (again, Kotlin, so apologies if it doesn't look like code that most people here would be used to):

https://github.com/varabyte/kobweb-site/commit/83a402ab5ca9198400354c90a5e8d02016ee5047

The relevant part is this that we've taken over rendering the component with Preact explicitly, no longer calling [`docsearch`](https://github.com/algolia/docsearch/blob/c591f004423a9ead953409f3d4a89643fa84b994/packages/docsearch-js/src/docsearch.tsx#L18), and force a `null` render call:

```kotlin
///////////// Before
Div(SearchStyle.toAttrs()) {
   DisposableEffect(Unit) {
      initAlgoliaSearch(scopeElement, ctx.router)
      onDispose { }
   }
}

private fun initAlgoliaSearch(element: HTMLElement, router: Router) {
   docsearch(
      json(
         "container" to element,
         "appId" to "...",
      )
   )
}

///////////// After
Div(SearchStyle.toAttrs()) {
   DisposableEffect(Unit) {
      initAlgoliaSearch(scopeElement, ctx.router)
      onDispose {
         Preact.render(null, scopeElement)
      }
   }
}

private fun initAlgoliaSearch(element: HTMLElement, router: Router) {
   Preact.render(
      Preact.h(
         DocSearchModule.DocSearch,
         json(
            "container" to element,
            "appId" to "...",
        )
   )
}
```

In other words, it feels like the React disposal logic for the `docsearch` function isn't set up correct? And as a result we suspect [this code](https://github.com/algolia/docsearch/blob/c591f004423a9ead953409f3d4a89643fa84b994/packages/docsearch-react/src/useDocSearchKeyboardEvents.ts#L60) isn't getting triggered.

**Comment by shortcuts** [comment]
Hey sorry all for the long silence here, i'm not working on docsearch anymore so I totally missed the notifications, I'll let the team now in charge look at this!

**Comment by 8bittitan** [comment]
@bitspittle @Mister-Hope I've left request for changes on the original [PR](https://github.com/algolia/docsearch/pull/1742) as it doesn't seem like the proposed solution would work.

**Comment by bitspittle** [comment]
In case @Mister-Hope doesn't ever reply, the collaborator I work with mentioned he'd also be OK if the docs for `docsearch` just mention you should dispose the call using Preact's `render(null, searchElement)` method.

In our (Kotlin) codebase ([link to code](https://github.com/varabyte/kobweb-site/blob/8320b85d2f8497cf4c1d92d361c4fda8d7df36d0/site/src/jsMain/kotlin/com/varabyte/kobweb/site/components/widgets/Search.kt#L93)), this effectively looks like:
```kotlin
DisposableEffect(Unit) {
   docsearch(
      json(
         "container" to searchElement,
         "appId" to "...",
         "apiKey" to "...",
      )
   )
   onDispose {
      // Release listeners hooked up by `docsearch`
      Preact.render(null, searchElement)
   }
}
```

**Comment by 8bittitan** [comment]
@bitspittle I think that's valid and we can add something to our docs to address it for now. Ultimately we should implement a solution in the library to handle this though.

**Comment by xav-ie** [comment]
In above PR, I try to address the issue. I am happy to take further feedback and fixes there, but I am pretty sure it works as is! :)))
